backend/src/worker.rs

```diff
@@ -1351,175 +1361,228 @@ async fn handle_child(
+            let mut read_lines = ready(line)
+                .into_stream()
+                .chain(output.by_ref())
+                .take_until(until)
+                .boxed();
+
+            /* Read up until an error is encountered,
+             * handle log lines first and then the error...
+             *
+             * (This looks a bit nicer using try_for_each but the side-effects/capturing
+             * in FnMut closure seems impractical, _maybe_ a futures::Sink would work but
+             * I know next to nothing about that.) */
+            let mut joined = String::new();
+
+            while let Some(line) = read_lines.next().await {
+                match line {
+                    Ok(_) if log_remaining == 0 => (),
+                    Ok(line) => {
+                        append_with_limit(&mut joined, &line, &mut log_remaining);
+
+                        *last_line = line;
+
+                        if log_remaining == 0 {
+                            tracing::info!(%job_id, "Too many logs lines");
+                            let _ = set_too_many_logs.send(true);
+                            joined.push_str(&format!(
+                                "Too many logs lines. Limit is {MAX_LOG_SIZE} chars. Killing job."
+                            ));
+                        }
                     }
-                },
-                Ok(Some(err)) = stderr_reader.next_line() => {
-                    if err.len() > MAX_LOG_SIZE as usize {
-                        tracing::info!("Line is too big");
-                        let _ = tx.send(format!("Line is too big")).await;
-                        done4.store(true, Ordering::Relaxed);
+                    Err(err) => {
+                        result = Err(err);
                         break;
-                    } else {
-                        tx.send(err).await
                     }
-                },
-                else => {
-                    break
-                },
-            };
-            if send.err().is_some() {
-                tracing::error!("error sending log line");
-            };
-        }
-    });
-
-    let db2 = db.clone();
-
-    tokio::spawn(async move {
-        while !&done3.load(Ordering::Relaxed) {
-            let q = sqlx::query!("UPDATE queue SET last_ping = now() WHERE id = $1", job_id)
-                .execute(&db2)
-                .await;
-
-            if q.is_err() {
-                tracing::error!("error setting last ping for id {}", job_id);
+                }
             }
 
-            tokio::time::sleep(Duration::from_secs(5)).await;
+            logs.push_str(&joined);
+
+            /* Ensure the last flush completed before starting a new one.
+             *
+             * We want append_logs() to run concurrently while reading output from the
+             * job process.  take_until() should only stop reading logs once either:
+             * - `write` has resovled
```

Comment:
I'm confused. Is append_logs really running concurrently in the sense that there is no new parallel thread spawned to run it ?

Reply:
Good catch. My understanding is that the future should run as long as it's being polled on, and it's being polled on in the `take_until()`. However, it's entirely possible that it doesn't even start that the scheduled write future until after `sleep(write_logs_delay)` is resolved.

Generally speaking something like:

```rust
future::join(sleep(Duration::from_secs(1)), sleep(Duration::from_secs(1))).await;
```

Will run both futures concurrently in the same task until they both finish and will resolve after one second. But _I think_ only because `join` will poll both of them. Whereas `sleep(...).then(bar)` will only start polling `bar` after the `sleep` resolves. I guess I could just `join(sleep(...), db_write_future)`.

Reply:
Although, even won't run under `while let Some(line) = output.by_ref().next().await {`. So it might be best as a task. If a task panics it it caught in a JoinHandle so it adds another thing to deal with but it may make this scheduling way easier to think about.

Reply:
I updated this to use a task. It's still a little ugly.